keyring.go

```diff
@@ -0,0 +1,303 @@
+// A JSON message is valid for a server if the message has at least one valid
+// signature from that server.
+type VerifyJSONRequest struct {
+	// The name of the matrix server to check for a signature for.
+	ServerName string
+	// The millisecond posix timestamp the message needs to be valid at.
+	AtTS Timestamp
+	// The JSON bytes.
+	Message []byte
+}
+
+// A VerifyJSONResult is the result of checking the signature of a JSON message.
+type VerifyJSONResult struct {
+	// Whether the message passed the signature checks.
+	// This will be nil if the message passed the checks.
+	// This will have an error if the message did not pass the checks.
+	Result error
+}
+
+// VerifyJSONs performs bulk JSON signature verification for a list of VerifyJSONRequests.
+// Returns a list of VerifyJSONResults with the same length and order as the request list.
+// The caller should check the Result field for each entry to see if it was valid.
+// Returns an error if there was a problem talking to the database or one of the other methods
+// of fetching the public keys.
+func (k *KeyRing) VerifyJSONs(requests []VerifyJSONRequest) ([]VerifyJSONResult, error) {
+	results := make([]VerifyJSONResult, len(requests))
+	keyIDs := make([][]string, len(requests))
+
+	for i := range requests {
+		ids, err := ListKeyIDs(requests[i].ServerName, requests[i].Message)
+		if err != nil {
+			results[i].Result = fmt.Errorf("gomatrixserverlib: error extracting key IDs")
+			continue
+		}
+		for _, keyID := range ids {
+			if k.isAlgorithmSupported(keyID) {
+				keyIDs[i] = append(keyIDs[i], keyID)
+			}
+		}
+		if len(keyIDs[i]) == 0 {
+			results[i].Result = fmt.Errorf(
+				"gomatrixserverlib: not signed by %q with a supported algorithm", requests[i].ServerName,
+			)
+			continue
+		}
+		// Set a place holder error in the result field.
+		// This will be unset if one of the signature checks passes.
+		// This will be overwritten if one of the signature checks fails.
+		// Therefore this will only remain in place if the keys couldn't be downloaded.
+		results[i].Result = fmt.Errorf(
+			"gomatrixserverlib: could not download key for %q", requests[i].ServerName,
+		)
+	}
+
+	keyRequests := k.publicKeyRequests(requests, results, keyIDs)
+	if len(keyRequests) == 0 {
+		// There aren't any keys to fetch so we can stop here.
+		// This will happen if all the objects are missing supported signatures.
+		return results, nil
+	}
+	keysFromDatabase, err := k.KeyDatabase.FetchKeys(keyRequests)
+	if err != nil {
+		return nil, err
+	}
+	k.checkUsingKeys(requests, results, keyIDs, keysFromDatabase)
+
+	for i := range k.KeyFetchers {
+		keyRequests := k.publicKeyRequests(requests, results, keyIDs)
+		if len(keyRequests) == 0 {
+			// There aren't any keys to fetch so we can stop here.
+			// This means that we've checked every JSON object we can check.
+			return results, nil
+		}
+		keysFetched, err := k.KeyFetchers[i].FetchKeys(keyRequests)
```

Comment:
Maybe, but the levels of synchronisation between goroutines, and possibly processes needed to prevent that is quite saddening. I've left a TODO e8de828